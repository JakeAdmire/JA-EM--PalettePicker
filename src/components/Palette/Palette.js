import React, { Component } from 'react';
import { Color } from '../Color/Color';
import { fetchProjects } from '../../utils/api';

export class Palette extends Component {
  constructor() {
    super();
    this.state = { 
      colors: [],
      savedProjects: [],
      savedPalettes: [],
    };
  }

  componentDidMount = async () => {
    this.regeneratePalette();
    document.addEventListener('keydown', this.handleKeydown);
    this.fetchProjects()
    this.fetchPalettes()
  }

  fetchProjects = async () => {
    let projects = await fetchApiData('projects')
    this.setState({
      savedProjects: projects[0]
    })
  }

  fetchPalettes = async () => {
    let palettes = await fetchApiData('palettes')
    this.setState({
      savedPalettes: palettes[0]
    })
  }

  handleKeydown = (event) => {
    if (event.keyCode === 32) this.regeneratePalette();
  }

  regeneratePalette = (colors = []) => {
    for (let i = 0; i < 5; i++) {
      let randColor = this.generateColor();
      colors.push(randColor);
    }
    this.setState({ colors });
  }

  generateColor() {
    return '#' + Math.random().toString(16).slice(2, 8).toUpperCase();
  }

  render() {
    const { colors } = this.state;

    let paletteClass = this.props.active ? 'active' : '';

    return colors 
      ? <div className={`Palette ${paletteClass}`}>
        <div className="colors">
        {
          colors.map(color => <Color key={color} hex={color}/>) 
        }
        </div>
        <div className={`sidebar`}>sidebar{this.state.savedProjects.name}</div>
      </div>
      : <p>LOADING..</p>
  }
}