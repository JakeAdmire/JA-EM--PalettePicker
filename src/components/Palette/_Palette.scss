.Palette {
  width: 100%; height: 100%;
  box-sizing: border-box;

  position: relative;
  
  display: flex;

  overflow: hidden;

  transition: all .2s ease;

  .colors {
    width: 100%;
    height: 100%;
    position: relative;
    display: flex;
  }

  .sidebar {
    background: grey;
    width: 240px; height: 100%;

    position: absolute;
    right: -240px;

    transition: all .2s ease;
  }
}

.active {
  padding-right: 240px;
}