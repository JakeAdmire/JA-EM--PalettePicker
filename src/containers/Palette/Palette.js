import React, { Component } from 'react';
import { Color } from '../../components/Color/Color';
import { currentColors } from '../../actions';
import { connect } from 'react-redux';

export class Palette extends Component {
  constructor() {
    super();
    this.state = {  colors: [] };
  }

  componentDidMount = async () => {
    this.regeneratePalette();
    document.addEventListener('keydown', this.handleKeydown);
  }

  handleKeydown = (event) => {
    if (event.keyCode === 32) this.regeneratePalette(this.state.colors);
  }

  regeneratePalette = (colors = []) => {
    let initialColors = [];

    for (let i = 0; i < 5; i++) {

      if (!colors[i] || !colors[i].locked) {
        initialColors.push({ hex: this.generateColor(), locked: false });

      } else if (colors[i].locked) {
        initialColors.push(colors[i]);
      }

    }
    this.setState({ colors: initialColors });
    this.props.currentColors(initialColors)
  }

  generateColor() {
    return '#' + Math.random().toString(16).slice(2, 8).toUpperCase();
  }

  toggleLocked = (hex) => {
    let storedPalette = this.state.colors;

    let foundColor = storedPalette.find(color => color.hex === hex)
    foundColor.locked = !foundColor.locked;

    this.setState({ colors: storedPalette });
  }

  render() {
    const { colors } = this.state;

    return colors 
      ? <div className={`Palette`}>
        <div className="colors">
        {
          colors.map(color => <Color key={color.hex} toggleLocked={this.toggleLocked} {...color}/>) 
        }
        </div>
      </div>
      : <p>PRESS SPACE TO GENERATE A NEW COLOR PALETTE!</p>
  }
}

export const mapDispatchToProps = (dispatch) => ({
  currentColors: (colors) => dispatch(currentColors(colors))
})

export default connect(null, mapDispatchToProps)(Palette)