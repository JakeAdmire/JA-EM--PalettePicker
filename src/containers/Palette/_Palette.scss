.Palette {
  width: 100%; height: 100%;
  box-sizing: border-box;

  position: relative;
  
  display: flex;

  transition: all .2s ease;

  .colors {
    width: 100%;
    height: 100%;
    position: relative;
    display: flex;
  }
}