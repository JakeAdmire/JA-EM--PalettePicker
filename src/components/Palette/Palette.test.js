import React from 'react';
import ReactDOM from 'react-dom';
import { Palette } from './Palette';
import { shallow } from 'enzyme';

describe('Palette', () => {

  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <Palette />
    )
  })

  it('matches the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should have default state', () => {
    let mockState = { colors: [] };
    expect(wrapper.state()).toEqual(mockState);
  })

  describe('handleKeydown(event)', () => {

    it.skip('should be invoked on any key press', () => {
      const mockSpy = jest.spyOn(wrapper.instance(), 'handleKeydown');
      const keyCode = { keyCode: 13 };

      wrapper.simulate('keyown', keyCode);

      expect(mockSpy).toHaveBeenCalledWith(keyCode);
    })

    it('should invoke generatePalette on space key press', () => {
      const mockEvent = { keyCode: 32 };
      const mockSpy = jest.spyOn(wrapper.instance(), 'generatePalette');

      wrapper.instance().handleKeydown(mockEvent);

      expect(mockSpy).toHaveBeenCalled();
    })

  })

  describe('generatePalette()', () => {

    it('should update state', () => {
      expect(wrapper.state('colors').length).toEqual(0)

      wrapper.instance().generatePalette();

      expect(wrapper.state('colors').length).toEqual(5)
    })

    it('should invoke generateBaseColor', () => {
      expect(wrapper.state('colors')).toEqual([]);
      const mockSpy = jest.spyOn(wrapper.instance(), 'generateBaseColor');

      wrapper.instance().generatePalette();

      expect(mockSpy).toHaveBeenCalled();
    })

    it('should invoke generateColor', () => {
      const mockSpy = jest.spyOn(wrapper.instance(), 'generateColor');

      wrapper.instance().generatePalette();

      expect(mockSpy).toHaveBeenCalled();
    })

  })

  describe('generateBaseColor()', () => {

    it('should return an HSL array with 3 elements', () => {
      const results = wrapper.instance().generateBaseColor();

      expect(results.length).toEqual(3);
    })

  })

  describe('generateColor(baseHSL, prevHSL)', () => {

    it('should return an HSL array based on its parameters', () => {
      let baseHSL = [225, 33, 59];
      let prevHSL = [338, 29, 63];

      let hslColor = wrapper.instance().generateColor(baseHSL, prevHSL);

      expect(hslColor.length).toEqual(3);
    })

  })

  describe('numberGenerator(min, max)', () => {

    it('should return a random number within the given parameters', () => {
      let mockMin = 0;
      let mockMax = 0;

      const results = wrapper.instance().numberGenerator(mockMin, mockMax);

      expect(results).toBeGreaterThanOrEqual(mockMin);
      expect(results).toBeLessThanOrEqual(mockMax);
    })

  })

  describe('toggleLocked(hex)', () => {

    it('should update state', () => {
      wrapper.instance().generatePalette();
      const initialState = wrapper.state('colors');

      wrapper.instance().toggleLocked(initialState[1].hex);

      expect(wrapper.state('colors')).toEqual(initialState);
    })

  })

})