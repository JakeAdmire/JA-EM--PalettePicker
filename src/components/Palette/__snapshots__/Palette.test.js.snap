// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Palette matches the snapshot 1`] = `ShallowWrapper {}`;
